nixos/recyclarr: init module

omares opened the pull request:
Adds a service definition that allows to run [Recyclarr](https://recyclarr.dev/wiki/) ([nixos packge](https://search.nixos.org/packages?channel=unstable&show=recyclarr&from=0&size=50&sort=alpha_asc&type=packages&query=recyclarr)) periodically.

Executing the sync regularly ensures that the Sonarr and Radarr instances stay updated with the latest configurations from the TRaSH guides

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

omares committed:
nixos/recyclarr: init module

ambroisie commented on nixos/modules/services/misc/recyclarr.nix:
```diff
@@ -0,0 +1,141 @@
+  ...
+}:
+
+with lib;
```
We're trying to reducer usage of `with lib;` in NixOS.

IMO, having it over `options` and `meta` is fine, but people are pushing to remove it completely (see #340346 and #208242).

ambroisie commented on nixos/modules/services/misc/recyclarr.nix:
```diff
@@ -0,0 +1,141 @@
+  options.services.recyclarr = {
+    enable = mkEnableOption "recyclarr service";
+
+    package = mkOption {
```
Use `mkPackageOption`.

ambroisie commented on nixos/modules/services/misc/recyclarr.nix:
```diff
@@ -0,0 +1,141 @@
+      description = "The recyclarr command to run (e.g., sync).";
+    };
+
+    configFile = mkOption {
```
Use RFC-42-style `settings` please. If you're worried about secrets, I'd recommend using `utils.genJqSecretsReplacementSnippet`.

omares replied:
Would the expectation be that a YAML file is generated based on the given settings, and its path is passed as the `--config` argument to the binary?

ambroisie replied:
Yeah, you can look at the PR I made to `nginx-sso` as an example.

omares replied:
With this approach, the file generation is moved into the service and out of the user's hands, making it cumbersome to use tools like SOPS or similar that are way better at managing secrets. Are we sure this is the correct solution in this case?

ambroisie replied:
That's the opposite: using `utils.genJqSecretsReplacementSnippet` makes it very easy to integrate with agenix/sops-nix to both set the configuration in nix while leaving secret handling to those tools.

I'm not sure I follow what is troubling you about it.

omares replied:
The lack of documentation regarding the use of `genJqSecretsReplacementSnippet` made it unclear how it works. I will apply the change and conduct a test usage.

ambroisie replied:
Yeah, not sure why its documentation isn't surfaced in the manual/better documented. If the current usage in other modules isn't enough "documentation", I can walk you through its usage

omares replied:
I applied the change and verified that the service is still working. I had to loosen some security permissions, but nothing major.

ambroisie commented on nixos/modules/services/misc/recyclarr.nix:
```diff
@@ -0,0 +1,141 @@
+      '';
+    };
+
+    workingDir = mkOption {
```
Is there a use case for changing it?

omares replied:
If not specified, the recyclarr executable will use the directory where the binary file is located as the working directory. I assumed that if a specify location is defined, users should have the option to change it.

ambroisie commented on nixos/modules/services/misc/recyclarr.nix:
```diff
@@ -0,0 +1,141 @@
+      description = "Working directory for recyclarr.";
+    };
+
+    schedule = mkOption {
```
I believe there is a type for systemd timer strings, though I can't remember it nor look it up on my phone.

ambroisie commented on nixos/modules/services/misc/recyclarr.nix:
```diff
@@ -0,0 +1,141 @@
+
+    schedule = mkOption {
+      type = types.str;
+      default = "*-*-* 04:00:00"; # Every day at 4 AM
```
I'd use `daily` for an easy default instead of the cron-style way.

ambroisie replied:
Also, it probably could use the random delay, persistence, etc... options that usually go with a timer.

omares replied:
Had those in my `systemd.timers` definition below; I will move them to the option.

omares replied:
After searching for examples, I only found [restic](https://github.com/NixOS/nixpkgs/blob/b8228735c965c9dd1356bb8725a34c24b2574844/nixos/modules/services/backup/restic.nix#L141) pursuing this approach. Other services typically expect just the interval, for simplicity, I would suggest keeping the current option definition. WDYT?

ambroisie replied:
Sounds good to me.

ambroisie commented on nixos/modules/services/misc/recyclarr.nix:
```diff
@@ -0,0 +1,141 @@
+      ${cfg.group} = { };
+    };
+
+    systemd.tmpfiles.rules = [
```
Use `systemd.tmpfiles.settings`.

ambroisie commented on nixos/modules/services/misc/recyclarr.nix:
```diff
@@ -0,0 +1,141 @@
+
+    systemd.services.recyclarr = {
+      description = "Recyclarr Service";
+      after = [ "network.target" ];
```
I think this isn't necessary nor really useful.

ambroisie commented on nixos/modules/services/misc/recyclarr.nix:
```diff
@@ -0,0 +1,141 @@
+    systemd.services.recyclarr = {
+      description = "Recyclarr Service";
+      after = [ "network.target" ];
+      path = [ cfg.package ];
```
Is this needed?

omares replied:
No, it was remnant before using `lib.getExe`.

ambroisie reviewed (commented):
Reviewed on my phone, so kind of terse.

Can't `recyclarr` get away with running as a `DynamicUser=true` service? It shouldn't interact with many on-disk files unless I'm missing something?

omares commented:
> Reviewed on my phone, so kind of terse.
> 
> Can't `recyclarr` get away with running as a `DynamicUser=true` service? It shouldn't interact with many on-disk files unless I'm missing something?

In theory, that should be possible, but I am not a fan of being unable to pin the UID and GID to ensure consistency across multiple system when needed and therefore tend to use specific user and group options.

ambroisie commented:
> In theory, that should be possible, but I am not a fan of being unable to pin the UID and GID to ensure consistency across multiple system when needed and therefore tend to use specific user and group options.

IMO this is a downstream consumer thing you could handle in your own configuration, though I don't have very strong feelings about it.

omares commented:
@ambroisie, I would appreciate it if you could provide me with another review and, ideally, your approval. Thank you!

ambroisie commented on nixos/modules/services/misc/recyclarr.nix:
````diff
@@ -0,0 +1,166 @@
+
+        The configuration is processed using [utils.genJqSecretsReplacementSnippet](https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/utils.nix#L232-L331) to handle secret substitution.
+
+        To avoid permission issues, secrets should be provided via systemd's credential mechanism:
+
+        ```nix
+        systemd.services.recyclarr.serviceConfig.LoadCredential = [
+          "radarr-api_key:''${config.sops.secrets.radarr-api_key.path}"
+        ];
````
Oh that's a pretty nice way to handle this.

I recall @Stunkymonkey mentioned wanting to work towards a reusable and automated way to do this (`LoadCredential`) for any module needing secrets.

@Stunkymonkey this seems relevant to your interests, if you also want to review it.

ambroisie commented on nixos/modules/services/misc/recyclarr.nix:
```diff
@@ -0,0 +1,166 @@
+let
+  cfg = config.services.recyclarr;
+  format = pkgs.formats.yaml { };
+  configPath = "/var/lib/recyclarr/config.json";
```
See, this is where I'm wondering whether having `cfg.workingDir` is useful: it makes handling the configuration with secrets more compilcated (where do we store it, in `workingDir` or in a `StateDirectory`? What if they're the same? etc...)

At the very least, given that you don't have a `StateDirectory` in your service, this path isn't guaranteed to exist.

IMO I'd just remove `workingDir`, use `StateDirectory = "recyclarr"` (which allows us to remove the `tempfiles` rule) and tell it to use `/var/lib/recyclarr`.

If you have strong feelings against it, please at least handle the configuration path correctly for non-default `workingDir` values.

omares replied:
Removed `workingDir` in favor of `StateDirectory` approach.

ambroisie commented on nixos/modules/services/misc/recyclarr.nix:
```diff
@@ -0,0 +1,166 @@
+    systemd.services.recyclarr = {
+      description = "Recyclarr Service";
+
+      preStart = utils.genJqSecretsReplacementSnippet cfg.configuration configPath;
```
An explicit note about using JSON instead of YAML makes it easier to know it isn't a mistake when reading it later.

```suggestion
      # YAML is a JSON super-set
      preStart = utils.genJqSecretsReplacementSnippet cfg.configuration configPath;
```

ambroisie reviewed (commented):
A small nit about documentation, and a note about `configPath`/`workingDir`.

omares commented on nixos/modules/services/misc/recyclarr.nix:
```diff
@@ -0,0 +1,166 @@
+            };
+          }];
+        }
+      '';
```
Done

omares commented:
> A small nit about documentation, and a note about `configPath`/`workingDir`.

Thank you, addressed your points. Would you mind resolving the conversations when you see the points completed?

ambroisie commented:
> Would you mind resolving the conversations when you see the points completed?

Hmm you can't resolve them? Done.

omares commented:
> > Would you mind resolving the conversations when you see the points completed?
> 
> Hmm you can't resolve them? Done.

Oh, I could, but for me, it's good etiquette to let the person who provided the feedback resolve the comments to confirm that their feedback is addressed.

After approval, what are the next steps to merge the change?

ambroisie commented:
> After approval, what are the next steps to merge the change?

I'm leaving a bit of time for (eventual) other reviewers to chime in. If nobody screams, I'll merge it soon.

You can ping me if I forget to do so :-)

omares commented:
@ambroisie Do we want to proceed with merging?, I doubt that other people will contribute with a review.

ambroisie commented:
SGTM, merged.

undefined-landmark commented:
Hi, thanks for your effort in creating this module! Just wondering if it's not necessary to add a couple things:
- Add recylarr to `nixos/modules/module-list.nix`
- Add tests
- Add release notes

ambroisie commented:
You know, I don't know how I missed the module list and release notes @undefined-landmark 🙃.

As for the test, I'd expect it to be difficult to get it working in a NixOS test given that one needs to have *arr API keys (so pre-setup) in the configuration.

Can you open a PR for the missing bits, I'll approve it.

undefined-landmark commented:
I opened a PR for the release notes and `module-list.nix`: #380084

undefined-landmark commented:
The module contained `lib.mdDoc` which is deprecated. This caused a fail in the checks. As far as I could gather option descriptions are seen as markdown by default now (https://github.com/NixOS/nixpkgs/blob/a2a0316319cc0625ab75baba5084aacb272ba6d0/lib/options.nix#L407), so removing `lib.mdDoc` should not cause any issues in this case. While editing the code I also noticed some back ticks were missing in the description code block, added those as well :).

In short, my PR can be reviewed.
